### ISSUE TITLE
Receiving thousands of heart emojis causes Signal (and iOS) to crash

### ISSUE THREAD

**Issue by briankung** [WhisperSystems/Signal-iOS#2551]
- [x] I have searched open and closed issues for duplicates

----------------------------------------

### Bug description
Receiving thousands of heart emojis (6,480, to be exact) causes Signal and iOS to crash

### Steps to reproduce
- Receive 6,480 heart emojis (a [gist](https://gist.githubusercontent.com/briankung/8b6f56b04f13971b5c25b6b8744b8637/raw/3e56ab827343f8fe76c1ec697d611ae8607b898e/Signal%2520%25F0%259F%2592%2594%2520%25E2%2598%25A0.md) for your convenience) on an iOS Signal client via secure Signal chat

**Actual result:** The phone crashes. My wife had to force restart her phone to make it functional _as a phone_ (it was just a black screen). Signal crashed again when she tried to open it, but didn't take her entire phone with it. She managed to get to the main screen (list of recent chats) without crashing after she updated Signal, but the client either crashed or was extremely slow whenever she tried to open our chat window. ~~And it never successfully displayed all my hearts.~~

**Expected result:** Signal displays 6,480 hearts, or clips the message either when sending or receiving the message.

### Screenshots
This is after she updated and after fighting through loading screens and crashes:
https://user-images.githubusercontent.com/2836167/30572350-a5780b12-9cb2-11e7-8329-3dd5f3c61854.png

Oh! We found the hearts:
https://user-images.githubusercontent.com/2836167/30572887-17dd10aa-9cb6-11e7-83e6-2664a01554a8.PNG

### Device info

**Device**: iPhone 6 64GB
**iOS version**: 10.2.1 (14D27)
**Signal version:** 2.16.0.20 (after update)

### Link to debug log
https://gist.github.com/anonymous/b34459cd050b99a5f7177fe6ce759085

**Comment by briankung** [comment]
Note that I was only able to send this reasonably smoothly from Signal Desktop. Maybe it's more of a Signal Desktop issue? I tried to send the hearts back to my Android device from my wife's iPhone and it just truncated the message and sent it as a `.TXT` attachment.

Sending 6,480 hearts from my 2013 Moto X was...difficult (submitted a crash report from there, too), but I managed and it seemed to do much the same thing as the desktop version. I'm surprised. I expected another `TXT` file on the iOS side after my Android client finished sending the hearts. Can't tell if my Android Signal client truncated the message though.
